Rename authors to staff

emilgoldsmith opened the pull request:
Resolves #330.

Again not a fun PR to review, but I have some things I want to point out I'd like some input on / help with:

- What name do you think we should use for the new author route? Initially I just renamed all occurrences of author to staff including the route, but I think `staff member` sounds much better in many instances including in the route, inputs? I did a regex right now to replace some of the strings where it said staff to staff member as well, I think I might actually also want to rename the components to StaffMember. And maybe afterwards we can do one of the as we go migrations of if we see places where staff sounds weird we change it.
- I fixed all the bugs I could find except for the SearchableSelector, and I'm guessing it's me that introduced the bug and that it wasn't there before, but I was wondering whether @Helw150 could be so nice as to look at it since it's his code and help me debug the problem?
- I left authors as it is when attached to an article, as in the future we hope to also add media contributors as another relationship with articles (illustrators / videographers), and that's the only place it still makes sense.
- I'd be grateful if you could try and help me skim through it and see if I missed somewhere (replaced it to staff in a place where it should've stayed author for example), and I'll also try taking another look myself, but of course look at the migrations 

Have fun!

emilgoldsmith commented on src/components/admin/form-components/SearchableSelector.jsx:
```diff
@@ -104,7 +104,7 @@ export default class SearchableSelector extends React.Component {
 
     const noObjectsMessage = (
       <span style={{ color: 'rgba(0, 0, 0, 0.3)' }}>
-        `No {this.props.mode} are currently assigned to this article`
```
@Helw150 I noticed that these backticks were showing up in the UI, I guess that was just you forgetting that it was html and not JS here right? Also I feel like the header looks a bit weird of the staff section no... it stands out in a bad way compared to the rest? I'm not sure if we should just try and replicate the Material UI label css for that?

Helw150 replied:
Oop yeah on the backticks. I need to throw in Snapshots ASAP because they are quick&easy and would have allowed us to catch unintended UI bugs like this when refactoring. Modularized code should still render the same snapshot, which I failed here!

Agree on the header - we should use [TextFieldLabel](https://github.com/mui-org/material-ui/blob/v0.17.1/src/TextField/TextFieldLabel.js) for consistency with the rest of our Material-UI usage.

emilgoldsmith commented:
Fixed the searchable selector bug, this should now hopefully pass CI. The only thing that is a bit weird and would require a minor refactor is that the mode of the `SearchBar` should be staff but the SearchableSelector mode should be Authors since in the ArticleController it's a semantic subset, but I think we can merge this in and either Fred can handle it in his PR or we can do it separately.

It'd be great if I could get reviews on this asap so we can merge it in before it creates too many merge conflicts since it's another huge rename and also renamed a lot of files @Helw150 @frederikbrinck

Helw150 commented on __tests__/end-to-end/jest-setup.js:
```diff
@@ -1,2 +1,2 @@
-const SIMPLE_TEST_TIMEOUT = (process.env.CIRCLECI ? 20 : 10) * 1000;
+const SIMPLE_TEST_TIMEOUT = (process.env.CIRCLECI ? 30 : 10) * 1000;
```
Why did you have to change timeout for the rename?

emilgoldsmith replied:
Well I was fixing E2E tests anyway, and I've been thinking of upping this anyway as our E2E tests sometime fail but if you rerun them they work, and I'm not sure if it's because of too small a timeout, so did a "2 birds with one stone" thing here, sorry ;)

Helw150 commented on src/components/admin/form-components/SearchableSelector.jsx:
```diff
@@ -133,7 +133,7 @@ SearchableSelector.propTypes = {
       name: PropTypes.string.isRequired,
     }),
-  ).isRequired,
+  ),
```
Based on discussions we have had before on "value" props, this should be required. What's the reason for this change? You mention a bug in SearchableSelector, but don't ever say what that bug is. Is this the fix for that?

emilgoldsmith replied:
No I don't think this was the bug, the bug was from a rename of some state where I had missed one some where. 

This and the other weird proptypes related changes (sorry) were proptypes errors I stumbled upon while testing my own code, I think most of them stem from my eslint redo, I think there was an error here because ArticleController passed in `null` when there weren't currently any authors, so I added a default value. But you're right, should I try and change it so that it's required and handling the default in the parent?

emilgoldsmith replied:
This one needs reply @Helw150

emilgoldsmith replied:
The main point is. Do we allow an empty Selector to be shown as a null or undefined, or do we enforce it should be an empty array? I guess we used empty string in the others, so maybe we should enforce empty array here as well

Helw150 replied:
I don't see where in article controller is passing in `null`. On initial load, it passes `[]` and in falcor ToState it passes `_.toArray(article.authors)` which returns `[]` if there are no authors.(actually this might be a fix that only exists on my branch? Not sure). Those are the only places it sets author state other than when `SearchableSelector` calls its `onUpdate`.

That means, if it is passing `null` it's getting null from `SearchableSelector` at some point which we would want to fix. I'm happy if you perform this fix in this branch or if you revert this change and leave it as a *noisy* problem to be fixed. This change just makes it a silent problem though.

emilgoldsmith replied:
Ahh no I think I remember. I think that possibly `article.authors` could be null, and then `_.toArray` was throwing errors because it couldn't convert `null` to be an array, that's my best guess anyway. I'm pretty sure there was an error I fixed there anyway. Feel free to test it on your branch

emilgoldsmith replied:
No that can't have been it, checked that behaviour

emilgoldsmith replied:
Okay maybe I just didn't have a reason lol, I'm gonna test reverting it now

Helw150 replied:
Yeah, I tested that behavior before my comment. toArray always returns an array. I think it the problematic functon is `update from 'react-addons-update' in `SearchableSelector`.

Helw150 commented on src/components/main/Staff.jsx:
```diff
@@ -0,0 +1,42 @@
+    photo: React.PropTypes.string,
+    articles: React.PropTypes.object,
+  }),
+};
```
General comment for renames like this. It looks like you used `mv` and then `git add` and `git rm` which when combined with the changes made git's diff engine unable to tell staff and authors are the same file. 

I'm pretty sure if you use `git mv` instead, it will make PR's like this a little bit easier to review.

emilgoldsmith replied:
Oh interesting, I wonder if that saves some extra meta data, I never knew of that command. But yeah I basically did what you said, except in effect I used `mv` and then `git add -A` which does both the add and rm.

emilgoldsmith replied:
https://stackoverflow.com/a/1094392/5711883 ahh, I'm afraid according to this it can't be avoided (though locally I use a better diff algorithm so maybe we can do something similar with query params for PR reviews?)

Helw150 replied:
I think all `git mv` does is adds the changes to the path to the index, but not the changes to the file itself. It might actually still have the same problems here if you did the move and the added changes in the same commit, but if you do them as two separate commits - one becomes the rename - which git understands - and the other becomes the file changes. 

Not sure of the behavior if you commit both the changes and the rename in the same commit, but either way `git mv` is not much more effort than just `mv` 😄

emilgoldsmith replied:
Yeah I think it may even have the same problems, because I think I did do the renames in different commits, and I just tried the query params and couldn't get it to work

emilgoldsmith replied:
Yeah Github just sucks lol, check this out https://github.com/thegazelle-ad/gazelle-server/pull/349/commits/2b2d37f0e271e1000d49e9f6762ab89e90974759

emilgoldsmith replied:
For next time I can try to tell you guys the ranges from which actual edits were so you can check it in chunks on each side of the file rename

Helw150 replied:
Yeah, it's not a big deal since it's pretty easy to understand as long as I am doing an actual careful scan. It's just a pain that I thought `git mv` might fix, but I guess not haha.

Helw150 commented on src/components/main/TeamMembersList.jsx:
```diff
@@ -40,7 +41,7 @@ TeamMembersList.propTypes = {
       image_url: React.PropTypes.string,
       name: React.PropTypes.string.isRequired,
-      job_title: React.PropTypes.string.isRequired,
+      job_title: React.PropTypes.string,
```
If it's not required, shouldn't you be handling the above `||` logic with a PropTypes default value? For that matter, shouldn't this be failing linting since it has a non-required Prop without a default value?

emilgoldsmith replied:
I was wanting to do this! But I don't think you can set defaults for nested proptypes, actually haven't googled it, but I'll check now

emilgoldsmith replied:
https://github.com/facebook/react/issues/2568

Seems it won't happen. In this case I'm not sure if we should just disable the eslint rule about defaultProps and handle all defaults ourselves like we did before? I quite liked the linter enforcing defaults though

Helw150 replied:
I mean, this doesn't seem to actually be failing linting so it's fine. If it's not possible, it's not possible 😒

emilgoldsmith replied:
Yeah :(

Helw150 commented on src/components/main/TeamPageController.jsx:
```diff
@@ -30,7 +31,10 @@ export default class TeamPageController extends FalcorController {
         return <NotFound />;
       }
-      const teamData = this.state.data.semesters.latest;
+      const teamData = _.map(this.state.data.semesters.latest, team => ({
```
Not sure what this change is doing. I don't see you changing how `teamData` is used anywhere. Does this change stem from the rename?

I think as a general rule that would make Find and Replace PR's in general easier to review, it would be nice if all changes that require some sort of logic change and not just the straight up F&R were noted in the PR text.

emilgoldsmith replied:
Yeah sorry my bad, I just caught some prop types errors in the browser while testing since when I had done ESlint fixes I had changed some components to (in line with what we were doing with not accepting falcor pseudo-arrarys) to only accept arrays, but hadn't changed all the places that called it apparently. Will try and make that clearer another time!

emilgoldsmith commented:
My last 2 commits reverted the prop type changes to `SearchableSelector`, did some more renames because I keep confusing myself on exactly where I want to name stuff StaffMember but I think it's at least clear now, and if people want to change some variables to staff member as they pass files other times I'm okay with that. I also found a proptype error in the unpublish button that was probably also me that caused that in my eslint PR. 

I couldn't figure out creating an article through the mysql cli (or at least making it show in the admin interface) which makes me suspicious that I've broken the `ArticleList`, I'll create a new issue for that

emilgoldsmith commented:
@Helw150 when you see this feel free to merge this in for my sake if you find my last edits okay
